File: src/main/kotlin/g0201_0300/s0239_sliding_window_maximum/Solution.kt
package g0201_0300.s0239_sliding_window_maximum

// #Hard #Top_100_Liked_Questions #Top_Interview_Questions #Array #Heap_Priority_Queue
// #Sliding_Window #Queue #Monotonic_Queue #Udemy_Arrays #Big_O_Time_O(n*k)_Space_O(n+k)

import java.util.LinkedList

class Solution {
    fun maxSlidingWindow(nums: IntArray, k: Int): IntArray {
        val n = nums.size
        val res = IntArray(n - k + 1)
        var x = 0
        val dq = LinkedList<Int>()
        var i = 0
        var j = 0
        while (j < nums.size) {
            while (dq.isNotEmpty() && dq.peekLast() < nums[j]) {
                dq.pollLast()
            }
            dq.addLast(nums[j])
            if (j - i + 1 == k) {
                res[x] = dq.peekFirst()
                ++x
                if (dq.peekFirst() == nums[i]) {
                    dq.pollFirst()
                }
                ++i
            }
            ++j
        }
        return res
    }
}

File: src/main/scala/g0201_0300/s0239_sliding_window_maximum/Solution.scala
package g0201_0300.s0239_sliding_window_maximum

// #Hard #Top_100_Liked_Questions #Top_Interview_Questions #Array #Heap_Priority_Queue
// #Sliding_Window #Queue #Monotonic_Queue #Udemy_Arrays #Big_O_Time_O(n*k)_Space_O(n+k)
// #2023_11_07_Time_1149_ms_(86.67%)_Space_83.5_MB_(46.67%)

object Solution {
    def maxSlidingWindow(nums: Array[Int], k: Int): Array[Int] = {
        val q = new java.util.ArrayDeque[(Int, Int)]()

        def populate(num: Int, index: Int): Unit = {
            while (!q.isEmpty && q.peekLast()._2 + k <= index) {
                q.pollLast()
            }
            while (!q.isEmpty && (q.peekFirst()._1 <= num)) {
                q.pollFirst()
            }
            q.addFirst((num, index))
        }

        if (k == 1) {
            nums
        } else {
            for (i <- (0 until k - 1)) {
                populate(nums(i), i)
            }
            ((k - 1) until nums.size).map { i =>
                populate(nums(i), i)
                q.peekLast()._1
            }.toArray
        }
    }
}
